### PR TITLE
Mark Uploaded Keys with batch tag ids

### PR DESCRIPTION
**Pull request by EugenM-SAP**
## Checklist

* [x] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [x] Make sure `mvn install` runs for the whole project and, if you touched any code in the respective service, submission and distribution service can be run with `spring-boot:run`

## Description

The main goal of the changes in this PR is to mark diagnosis keys with the batch tag id to which they were associated during upload. To achieve this : 
- A new FederationUploadKey entity was introduced which contains the required extra metadata ( plus sql migration scripts)
- A new method to update the batchTagId field of the upload key in the FederationUploadService
- Refactoring of old upload code to use the new entity
- Additional unit tests and integration test

### CODE REVIEW

File: services/upload/src/test/java/app/coronawarn/server/services/federation/upload/integration/DiagnosisKeyUploadIT.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    Iterable<FederationUploadKey> currentKeys = uploadKeyRepository.findAll();
+
+    currentKeys.forEach(this::assertKeyWasMarkedWithBatchTagId);
+  }
+
+  private void assertKeyWasMarkedWithBatchTagId(FederationUploadKey uploadKey) {
+    Assertions.assertNotNull(uploadKey.getBatchTagId());
+  }
+
```
Comment (michael-burwig) [inline services/upload/src/test/java/app/coronawarn/server/services/federation/upload/integration/DiagnosisKeyUploadIT.java:74]: inconsistent padding

File: services/upload/src/main/java/app/coronawarn/server/services/federation/upload/payload/UploadPayload.java [hunk 2]
```diff
@@ -64,4 +69,16 @@ public UploadPayload setBatchTag(String batchTag) {
...
+   * Returns {@link DiagnosisKey} entities that are part of this upload payload.
+   */
+  public List<FederationUploadKey> getOriginalKeys() {
+    return originalKeys;
+  }
+
+  public void setOriginalKeys(List<FederationUploadKey> originalKeys) {
+    this.originalKeys = originalKeys;
+  }
+
```
Comment (michael-burwig) [inline services/upload/src/main/java/app/coronawarn/server/services/federation/upload/payload/UploadPayload.java:83]: inconsistent padding

File: services/upload/src/test/java/app/coronawarn/server/services/federation/upload/payload/helper/PersistenceKeysGenerator.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    byte[] bytes = new byte[16];
+    RANDOM.nextBytes(bytes);
+    return FederationUploadKey.from(
+        DiagnosisKey.builder()
+        .withKeyData(bytes)
+        .withRollingStartIntervalNumber(144)
+        .withTransmissionRiskLevel(0)
+        .build());
+  }
+
```
Comment (michael-burwig) [inline services/upload/src/test/java/app/coronawarn/server/services/federation/upload/payload/helper/PersistenceKeysGenerator.java:41]: inconsistent padding

File: services/upload/src/test/java/app/coronawarn/server/services/federation/upload/utils/MockData.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+public class MockData {
+
+  public static final String TEST_ORIGIN_COUNTRY = "DE";
+
+  public static List<FederationUploadKey> generateRandomUploadKeys(boolean consentToShare, int numberOfKeys) {
+    List<FederationUploadKey> fakeKeys = new ArrayList<>();
+    while (numberOfKeys > 0) {
+      fakeKeys.add(generateRandomUploadKey(consentToShare));
+      numberOfKeys--;
```
Comment (michael-burwig) [inline services/upload/src/test/java/app/coronawarn/server/services/federation/upload/utils/MockData.java:38]: Avoid mutating method arguments. Use `for`-loop or `IntStream.range(0, numberOfKeys)` instead.

File: services/upload/src/test/java/app/coronawarn/server/services/federation/upload/utils/MockData.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+   return FederationUploadKey.from(generateRandomDiagnosisKey(consentToShare));
+  }
+
+  public static List<DiagnosisKey> generateRandomDiagnosisKeys(boolean consentToShare, int numberOfKeys) {
+    List<DiagnosisKey> fakeKeys = new ArrayList<>();
+    while (numberOfKeys > 0) {
+      fakeKeys.add(generateRandomDiagnosisKey(consentToShare));
+      numberOfKeys--;
+    }
+    return fakeKeys;
```
Comment (michael-burwig) [inline services/upload/src/test/java/app/coronawarn/server/services/federation/upload/utils/MockData.java:53]:
```suggestion
    return IntStream.range(0, numberOfKeys)
      .mapToObj(ignore -> generateRandomDiagnosisKey(consentToShare))
      .collect(Collectors.toList());
```

File: services/upload/src/test/java/app/coronawarn/server/services/federation/upload/utils/MockData.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  public static DiagnosisKey generateRandomDiagnosisKey(boolean consentToShare) {
+     return DiagnosisKey.builder().withKeyData(randomByteData())
+                            .withRollingStartIntervalNumber(1)
```
Comment (michael-burwig) [inline services/upload/src/test/java/app/coronawarn/server/services/federation/upload/utils/MockData.java:58]: inconsistent indentation

File: services/upload/src/test/java/app/coronawarn/server/services/federation/upload/utils/PersistenceMockData.java [hunk 7]
```diff
@@ -24,23 +24,25 @@
...
+import app.coronawarn.server.common.persistence.domain.FederationUploadKey;
 import app.coronawarn.server.common.protocols.external.exposurenotification.ReportType;
 
 public class PersistenceMockData {
 
   public static final String TEST_ORIGIN_COUNTRY = "DE";
 
-  public static List<DiagnosisKey> generateRandomDiagnosisKeys(boolean consentToShare, int numberOfKeys) {
-    List<DiagnosisKey> fakeKeys = new ArrayList<DiagnosisKey>();
+  public static List<FederationUploadKey> generateRandomDiagnosisKeys(boolean consentToShare, int numberOfKeys) {
```
Comment (michael-burwig) [inline services/upload/src/test/java/app/coronawarn/server/services/federation/upload/utils/PersistenceMockData.java:34]: refrain from mutating method arguments

File: services/upload/src/test/resources/application-integration-test.yaml [hunk 8]
```diff
@@ -9,6 +9,13 @@ spring:
   flyway:
     password: local_setup_flyway
     user: local_setup_flyway
+    baseline-on-migrate: true
```
Comment (michael-burwig) [inline services/upload/src/test/resources/application-integration-test.yaml:12]: Why do we not need this in any other test?

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/domain/FederationUploadKey.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      ReportType reportType, int daysSinceOnsetOfSymptoms) {
+    super(keyData, rollingStartIntervalNumber, rollingPeriod, transmissionRiskLevel, submissionTimestamp,
+        consentToFederation, originCountry, visitedCountries, reportType, daysSinceOnsetOfSymptoms);
+  }
+
+  public String getBatchTagId() {
+    return batchTagId;
+  }
+
+  public void setBatchTagId(String batchTagId) {
```
Comment (michael-burwig) [inline common/persistence/src/main/java/app/coronawarn/server/common/persistence/domain/FederationUploadKey.java:50]: does not seem to be called

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/domain/FederationUploadKey.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * This entity is mapped to a table which serves as data source for the
+ * uploading of diagnosis keys to the Federation Gateway Service. The underlying table
+ * is filled by replicating diagnosis keys on insert (currently via database triggers).
+ *
+ * <p>The entity is almost equivalent to the {@link DiagnosisKey} in content, but carries
+ * some additional specific metadata, and could diverge in the future.
+ */
+public class FederationUploadKey extends DiagnosisKey {
```
Comment (michael-burwig) [inline common/persistence/src/main/java/app/coronawarn/server/common/persistence/domain/FederationUploadKey.java:35]: Consider overriding `hashCode` + `equals`

File: services/upload/src/main/java/app/coronawarn/server/services/federation/upload/keys/DiagnosisKeyGenerator.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+@Profile("testdata")
+public class DiagnosisKeyGenerator implements DiagnosisKeyLoader {
+
+  private final Random random = new Random();
+
+  private static final Logger logger = LoggerFactory
+      .getLogger(DiagnosisKeyGenerator.class);
```
Comment (michael-burwig) [inline services/upload/src/main/java/app/coronawarn/server/services/federation/upload/keys/DiagnosisKeyGenerator.java:43]: can be merged with previous line

**Review by michael-burwig — CHANGES_REQUESTED** [review]

File: services/upload/src/main/java/app/coronawarn/server/services/federation/upload/payload/PayloadFactory.java [hunk 12]
```diff
@@ -79,11 +82,13 @@ private String generateBatchTag(int counter) {
...
-    return batches.stream()
-        .map(b -> Pair.of(batches.indexOf(b), b))
-        .map(this::mapToPayloadAndSign)
+  public List<UploadPayload> makePayloadList(List<FederationUploadKey> diagnosisKeys) {
+    Map<DiagnosisKeyBatch, List<FederationUploadKey>> batchesAndOriginalKeys = assembler
+        .assembleDiagnosisKeyBatch(diagnosisKeys);
+    AtomicInteger batchCounter = new AtomicInteger(0);
+
+    return batchesAndOriginalKeys.entrySet().stream()
+        .map(entry -> this.mapToPayloadAndSign(batchCounter.incrementAndGet(),entry.getKey(), entry.getValue()))
```
Comment (michael-burwig) [inline services/upload/src/main/java/app/coronawarn/server/services/federation/upload/payload/PayloadFactory.java:91]: missing whitespace

File: common/persistence/src/main/resources/db/migration/V10__addBatchTagIdForUploadKeyTable.sql [hunk 13]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE federation_upload_key
+    ADD batch_tag  VARCHAR(50);
```
Comment (hilmarf) [inline common/persistence/src/main/resources/db/migration/V10__addBatchTagIdForUploadKeyTable.sql:2]: ensure this is later changed to varchar(64) - https://github.com/eu-federation-gateway-service/efgs-federation-gateway/blob/master/src/main/resources/db/changelog/v000-create-diagnosis-key-entity-table.yml#L24

**Review by hilmarf — APPROVED** [review]

**Pull request merged**